roles/translator/src/template_receiver/mod.rs

```diff
@@ -0,0 +1,142 @@
+                .recv()
+                .await
+                .unwrap();
+
+        let self_mutex = Arc::new(Mutex::new(Self {
+            receiver: receiver.clone(),
+            sender: sender.clone(),
+            send_new_tp_to_negotiator,
+            send_new_ph_to_negotiator,
+        }));
+
+        let sv2_frame: StdFrame = PoolMessages::TemplateDistribution(
+            roles_logic_sv2::parsers::TemplateDistribution::CoinbaseOutputDataSize(
+                coinbase_output_max_additional_size,
+            ),
+        )
+        .try_into()
+        .unwrap();
+        Self::send(self_mutex.clone(), sv2_frame).await;
+        let cloned = self_mutex.clone();
+
+        tokio::task::spawn(Self::on_new_solution(cloned, solution_receiver));
+        Self::start_templates(self_mutex, token);
+    }
+
+    pub async fn send(self_: Arc<Mutex<Self>>, sv2_frame: StdFrame) {
+        let either_frame = sv2_frame.into();
+        let sender_to_tp = self_.safe_lock(|self_| self_.sender.clone()).unwrap();
+        match sender_to_tp.send(either_frame).await {
+            Ok(_) => (),
+            Err(e) => panic!("{:?}", e),
+        }
+    }
+
+    pub fn start_templates(self_mutex: Arc<Mutex<Self>>, token: u64) {
+        tokio::task::spawn(async move {
+            // Send CoinbaseOutputDataSize size to TP
+            loop {
+                // Receive Templates and SetPrevHash from TP to send to JN
+                let receiver = self_mutex
+                    .clone()
+                    .safe_lock(|s| s.receiver.clone())
+                    .unwrap();
+                let mut frame: StdFrame = receiver.recv().await.unwrap().try_into().unwrap();
+                let message_type = frame.get_header().unwrap().msg_type();
+                let payload = frame.payload();
+
+                let next_message_to_send =
+                    ParseServerTemplateDistributionMessages::handle_message_template_distribution(
+                        self_mutex.clone(),
+                        message_type,
+                        payload,
+                    );
+                match next_message_to_send {
+                    Ok(SendTo::None(m)) => match m {
+                        Some(TemplateDistribution::NewTemplate(m)) => {
+                            crate::upstream_sv2::upstream::IS_NEW_TEMPLATE_HANDLED
+                                .store(false, std::sync::atomic::Ordering::SeqCst);
+                            let sender = self_mutex
+                                .safe_lock(|s| s.send_new_tp_to_negotiator.clone())
+                                .unwrap();
+                            sender.send((m, token)).await.unwrap();
+                        }
+                        Some(TemplateDistribution::SetNewPrevHash(m)) => {
+                            while !crate::upstream_sv2::upstream::IS_NEW_TEMPLATE_HANDLED
+                                .load(std::sync::atomic::Ordering::SeqCst)
```

Comment:
This could use some logging too - I could see this getting into an infinite loop if IS_NEW_TEMPLATE_HANDLED never gets set to true.